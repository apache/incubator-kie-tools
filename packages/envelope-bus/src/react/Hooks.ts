/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ApiDefinition, MessageBusClient, NotificationPropertyNames, SubscriptionCallback } from "../api";
import { useEffect } from "react";

export function useSubscription<Api extends ApiDefinition<Api>, M extends NotificationPropertyNames<Api>>(
  bus: MessageBusClient<Api>,
  method: M,
  callback: SubscriptionCallback<Api, M>
) {
  useEffect(() => {
    const subscription = bus.subscribe(method, callback);
    return () => {
      bus.unsubscribe(method, subscription);
    };
  }, [bus, method, callback]);
}

export function useSubscriptionOnce<Api extends ApiDefinition<Api>, M extends NotificationPropertyNames<Api>>(
  bus: MessageBusClient<Api>,
  method: M,
  callback: SubscriptionCallback<Api, M>
) {
  useEffect(() => {
    let unsubscribed = false;

    const subscription = bus.subscribe(method, (...args) => {
      callback(...args);
      unsubscribed = true;
      bus.unsubscribe(method, subscription);
    });

    return () => {
      if (!unsubscribed) {
        bus.unsubscribe(method, subscription);
      }
    };
  }, [bus, method, callback]);
}
