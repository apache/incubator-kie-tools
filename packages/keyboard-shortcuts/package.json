{
  "name": "@kie-tooling-core/keyboard-shortcuts",
  "version": "0.0.0",
  "description": "",
  "license": "Apache-2.0",
  "files": [
    "dist",
    "src"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/kiegroup/kogito-tooling.git"
  },
  "dependencies": {
    "@kie-tooling-core/envelope-bus": "0.0.0",
    "@kie-tooling-core/operating-system": "0.0.0",
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  },
  "devDependencies": {
    "@kogito-tooling/build-env": "0.0.0"
  },
  "scripts": {
    "lint": "yarn run run-script-if --bool \"$(build-env global.build.lint)\" --then \"yarn eslint ./src --ext .ts,.tsx\"",
    "test": "yarn run run-script-if --bool \"$(build-env global.build.test)\" --then \"jest --silent --verbose --passWithNoTests\"",
    "build:dev": "rimraf dist && tsc -p tsconfig.dev.json",
    "build:prod": "yarn lint && rimraf dist && tsc -p tsconfig.prod.json && yarn test"
  },
  "babel": {
    "presets": [
      [
        "@babel/env",
        {
          "targets": {
            "node": "current"
          }
        }
      ],
      "@babel/react"
    ]
  }
}
