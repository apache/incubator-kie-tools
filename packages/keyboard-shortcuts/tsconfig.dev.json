{
  "extends": "../../tsconfig.dev.json",
  "compilerOptions": {
    "outDir": "dist"
  },
  "include": ["src"]
}
