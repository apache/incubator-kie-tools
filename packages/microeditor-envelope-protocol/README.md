MicroEditors Envelope Protocol
--

Defines the messaging protocol from outside the `iframe` encapsulating the `AppFormer.Editor`. Should be consumed by any channel that features a MicroEditor.