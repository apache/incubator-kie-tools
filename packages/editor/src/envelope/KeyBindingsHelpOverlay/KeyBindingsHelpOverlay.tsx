/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { useCallback, useEffect, useMemo, useState } from "react";
import {
  Text,
  TextContent,
  TextList,
  TextListItem,
  TextListItemVariants,
  TextListVariants,
  TextVariants,
} from "@patternfly/react-core/dist/js/components/Text";
import { Modal } from "@patternfly/react-core/dist/js/components/Modal";
import { KeyboardIcon } from "@patternfly/react-icons/dist/js/icons/keyboard-icon";
import { OperatingSystem } from "@kie-tooling-core/operating-system";
import { useKogitoEditorEnvelopeContext } from "../../api";
import { useEditorEnvelopeI18nContext } from "../i18n";

export function KeyBindingsHelpOverlay() {
  const [showing, setShowing] = useState(false);
  const envelopeContext = useKogitoEditorEnvelopeContext();
  const { i18n } = useEditorEnvelopeI18nContext();

  const toggle = useCallback(() => {
    setShowing(!showing);
  }, [showing]);

  const keyBindings = useMemo(() => {
    return removeDuplicatesByAttr(envelopeContext.services.keyboardShortcuts.registered(), "combination")
      .filter((k) => !k.opts?.hidden)
      .map((k) => {
        return {
          combination: handleMacOsCombination(k.combination, envelopeContext.operatingSystem),
          category: k.label.split("|")[0]?.trim(),
          label: k.label.split("|")[1]?.trim(),
        };
      })
      .reduce((lhs, rhs) => {
        if (!lhs.has(rhs.category)) {
          lhs.set(rhs.category, new Set([{ label: rhs.label, combination: rhs.combination }]));
        } else {
          lhs.get(rhs.category)!.add({ label: rhs.label, combination: rhs.combination });
        }
        return lhs;
      }, new Map<string, Set<{ label: string; combination: string }>>());
  }, [envelopeContext.services.keyboardShortcuts.registered()]);

  useEffect(() => {
    const id = envelopeContext.services.keyboardShortcuts.registerKeyPress(
      "shift+/",
      `${i18n.keyBindingsHelpOverlay.categories.help} | ${i18n.keyBindingsHelpOverlay.commands.showKeyboardOverlay}`,
      async () => setShowing(true),
      { element: window }
    );
    return () => envelopeContext.services.keyboardShortcuts.deregister(id);
  }, [i18n]);

  useEffect(() => {
    if (showing) {
      const id = envelopeContext.services.keyboardShortcuts.registerKeyPressOnce("esc", async () => setShowing(false), {
        element: window,
      });
      return () => envelopeContext.services.keyboardShortcuts.deregister(id);
    }
  }, [showing]);

  return (
    <>
      <div
        onClick={() => setShowing(!showing)}
        className={"kogito-tooling--keyboard-shortcuts kogito-tooling--keyboard-shortcuts-icon"}
        data-testid={"keyboard-shortcuts-help-overlay-icon"}
      >
        <KeyboardIcon />
      </div>

      <Modal
        appendTo={document.body}
        title={i18n.keyBindingsHelpOverlay.title}
        isOpen={showing}
        width={"60%"}
        onClose={toggle}
        data-testid={"keyboard-shortcuts-help-overlay"}
        className={"kogito-tooling--keyboard-shortcuts"}
      >
        <TextContent>
          <TextList component={TextListVariants.dl}>
            {Array.from(keyBindings.keys()).map((category) => (
              <React.Fragment key={category}>
                <Text component={TextVariants.h2}>{category}</Text>
                {Array.from(keyBindings.get(category)!).map((keyBinding) => (
                  <React.Fragment key={keyBinding.combination}>
                    <TextListItem component={TextListItemVariants.dt}>
                      {formatKeyBindingCombination(keyBinding.combination)}
                    </TextListItem>
                    <TextListItem component={TextListItemVariants.dd}>{keyBinding.label}</TextListItem>
                  </React.Fragment>
                ))}
              </React.Fragment>
            ))}
          </TextList>
        </TextContent>
      </Modal>
    </>
  );
}

function handleMacOsCombination(combination: string, os?: OperatingSystem) {
  if (os === OperatingSystem.MACOS) {
    return combination.replace("ctrl", "cmd");
  }

  return combination;
}

function removeDuplicatesByAttr<T>(myArr: T[], prop: keyof T) {
  return myArr.filter((obj, pos, arr) => {
    return arr.map((mapObj) => mapObj[prop]).indexOf(obj[prop]) === pos;
  });
}

function formatKeyBindingCombination(combination: string) {
  return combination
    .split("+")
    .map((w) => w.replace(/^\w/, (c) => c.toUpperCase()))
    .join(" + ");
}
