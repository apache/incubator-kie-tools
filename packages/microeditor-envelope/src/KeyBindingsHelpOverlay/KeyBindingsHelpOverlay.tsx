/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { useEffect, useState } from "react";
import { KeyboardShortcutsApi } from "../api/keyboardShortcuts";

export function KeyBindingsHelpOverlay(props: { keyboardShortcuts: KeyboardShortcutsApi }) {
  const [showing, setShowing] = useState(false);

  const keyBindings = removeDuplicatesByAttr(props.keyboardShortcuts.registered(), "combination")
    .filter(k => !k.opts?.hidden)
    .map(k => {
      console.info("Mapping: " + k.label);
      return {
        combination: k.combination,
        category: k.label.split("|")[0]?.trim(),
        label: k.label.split("|")[1]?.trim()
      };
    })
    .reduce((lhs, rhs) => {
      if (!lhs.has(rhs.category)) {
        lhs.set(rhs.category, new Set([{ label: rhs.label, combination: rhs.combination }]));
      } else {
        lhs.get(rhs.category)!.add({ label: rhs.label, combination: rhs.combination });
      }
      return lhs;
    }, new Map<string, Set<{ label: string; combination: string }>>());

  useEffect(() => {
    const id = props.keyboardShortcuts.registerKeyPress(
      "shift+/",
      "Help | Show keyboard shortcuts",
      async () => setShowing(true),
      { element: window }
    );
    return () => props.keyboardShortcuts.deregister(id);
  }, []);

  useEffect(() => {
    let id: number;
    if (showing) {
      id = props.keyboardShortcuts.registerKeyPressOnce("esc", async () => setShowing(false), { element: window });
    }
    return () => {
      if (showing) {
        props.keyboardShortcuts.deregister(id);
      }
    };
  }, [showing]);

  return (
    <>
      <div
        onClick={() => setShowing(!showing)}
        style={{
          userSelect: "none",
          zIndex: 999,
          right: 0,
          bottom: 0,
          position: "fixed",
          padding: "7px",
          display: "flex",
          alignItems: "center",
          justifyContent: "center",
          width: "35px",
          height: "35px",
          fontSize: "1.2em",
          cursor: "pointer"
        }}
      >
        <b>{showing ? "x" : "?"}</b>
      </div>
      {showing && (
        <div
          style={{
            userSelect: "none",
            zIndex: 998,
            top: 0,
            left: 0,
            position: "fixed",
            width: "100vw",
            height: "100vh",
            padding: "40px",
            backdropFilter: "blur(5px)",
            background: "#cacacaa6"
          }}
        >
          <h1>Keyboard shortcuts</h1>
          {Array.from(keyBindings.keys()).map(category => (
            <p key={category}>
              <h3>{category}</h3>
              {Array.from(keyBindings.get(category)!).map(keyBinding => (
                <div key={keyBinding.combination}>
                  <span
                    style={{
                      color: "black",
                      backgroundColor: "white",
                      borderRadius: "4px",
                      fontFamily: "monospace",
                      fontSize: "0.8em",
                      padding: "1px 5px 1px 5px",
                      marginRight: "4px"
                    }}
                  >
                    {formatKeyBindingCombination(keyBinding.combination)}
                  </span>
                  <span>{keyBinding.label}</span>
                </div>
              ))}
            </p>
          ))}
        </div>
      )}
    </>
  );
}

function removeDuplicatesByAttr<T>(myArr: T[], prop: keyof T) {
  return myArr.filter((obj, pos, arr) => {
    return arr.map(mapObj => mapObj[prop]).indexOf(obj[prop]) === pos;
  });
}

function formatKeyBindingCombination(combination: string) {
  return combination
    .split("+")
    .map(w => w.replace(/^\w/, c => c.toUpperCase()))
    .join(" + ");
}
