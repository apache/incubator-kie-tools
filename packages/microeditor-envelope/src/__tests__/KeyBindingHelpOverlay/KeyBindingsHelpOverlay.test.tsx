/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { KeyBindingsHelpOverlay } from "../../KeyBindingsHelpOverlay";
import { DefaultKeyboardShortcutsService } from "@kogito-tooling/keyboard-shortcuts";
import { ChannelType, OperatingSystem } from "@kogito-tooling/microeditor-envelope-protocol";
import { fireEvent, render } from "@testing-library/react";
import { DEFAULT_TESTING_ENVELOPE_CONTEXT, usingEnvelopeApi } from "../envelopeApiUtils";

describe("KeyBindingsHelpOverlay", () => {
  test("minimal setup", async () => {
    const context = { operatingSystem: OperatingSystem.WINDOWS, channel: ChannelType.DESKTOP };
    const keyboardShortcutsService = new DefaultKeyboardShortcutsService({ editorContext: context });
    keyboardShortcutsService.registerKeyPress("ctrl+c", "Copy", () => Promise.resolve(), {});

    const component = render(
      usingEnvelopeApi(<KeyBindingsHelpOverlay />, {
        context: context,
        services: {
          ...DEFAULT_TESTING_ENVELOPE_CONTEXT.services,
          keyboardShortcuts: keyboardShortcutsService,
        }
      }).wrapper
    );

    fireEvent.click(component.getByTestId("keyboard-shortcuts-help-overlay-icon"));

    await component.findByTestId("keyboard-shortcuts-help-overlay");
    await component.findByText("Copy");

    expect(component.baseElement).toMatchSnapshot();
  });
});
