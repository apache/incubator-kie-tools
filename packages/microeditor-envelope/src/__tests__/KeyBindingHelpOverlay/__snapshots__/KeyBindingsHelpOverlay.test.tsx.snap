// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeyBindingsHelpOverlay minimal setup 1`] = `
<body
  class="pf-c-backdrop__open"
>
  <div
    aria-hidden="true"
  >
    <div
      class="kogito-tooling--keyboard-shortcuts-icon"
      data-testid="keyboard-shortcuts-help-overlay-icon"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 576 512"
        width="1em"
      >
        <path
          d="M528 448H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h480c26.51 0 48 21.49 48 48v288c0 26.51-21.49 48-48 48zM128 180v-40c0-6.627-5.373-12-12-12H76c-6.627 0-12 5.373-12 12v40c0 6.627 5.373 12 12 12h40c6.627 0 12-5.373 12-12zm96 0v-40c0-6.627-5.373-12-12-12h-40c-6.627 0-12 5.373-12 12v40c0 6.627 5.373 12 12 12h40c6.627 0 12-5.373 12-12zm96 0v-40c0-6.627-5.373-12-12-12h-40c-6.627 0-12 5.373-12 12v40c0 6.627 5.373 12 12 12h40c6.627 0 12-5.373 12-12zm96 0v-40c0-6.627-5.373-12-12-12h-40c-6.627 0-12 5.373-12 12v40c0 6.627 5.373 12 12 12h40c6.627 0 12-5.373 12-12zm96 0v-40c0-6.627-5.373-12-12-12h-40c-6.627 0-12 5.373-12 12v40c0 6.627 5.373 12 12 12h40c6.627 0 12-5.373 12-12zm-336 96v-40c0-6.627-5.373-12-12-12h-40c-6.627 0-12 5.373-12 12v40c0 6.627 5.373 12 12 12h40c6.627 0 12-5.373 12-12zm96 0v-40c0-6.627-5.373-12-12-12h-40c-6.627 0-12 5.373-12 12v40c0 6.627 5.373 12 12 12h40c6.627 0 12-5.373 12-12zm96 0v-40c0-6.627-5.373-12-12-12h-40c-6.627 0-12 5.373-12 12v40c0 6.627 5.373 12 12 12h40c6.627 0 12-5.373 12-12zm96 0v-40c0-6.627-5.373-12-12-12h-40c-6.627 0-12 5.373-12 12v40c0 6.627 5.373 12 12 12h40c6.627 0 12-5.373 12-12zm-336 96v-40c0-6.627-5.373-12-12-12H76c-6.627 0-12 5.373-12 12v40c0 6.627 5.373 12 12 12h40c6.627 0 12-5.373 12-12zm288 0v-40c0-6.627-5.373-12-12-12H172c-6.627 0-12 5.373-12 12v40c0 6.627 5.373 12 12 12h232c6.627 0 12-5.373 12-12zm96 0v-40c0-6.627-5.373-12-12-12h-40c-6.627 0-12 5.373-12 12v40c0 6.627 5.373 12 12 12h40c6.627 0 12-5.373 12-12z"
          transform=""
        />
      </svg>
    </div>
  </div>
  <div>
    <div
      class="pf-c-backdrop"
    >
      <div
        class="pf-l-bullseye"
      >
        <div
          aria-describedby="pf-modal-0"
          aria-label="Keyboard shortcuts"
          aria-modal="true"
          class="pf-c-modal-box"
          role="dialog"
          style="width: 60%;"
        >
          <button
            aria-label="Close"
            class="pf-c-button pf-m-plain"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 352 512"
              width="1em"
            >
              <path
                d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                transform=""
              />
            </svg>
          </button>
          <h1
            class="pf-c-title pf-m-2xl"
          >
             
            Keyboard shortcuts
             
          </h1>
          <div
            class="pf-c-modal-box__body"
            data-testid="keyboard-shortcuts-help-overlay"
            id="pf-modal-0"
          >
            <div
              class="pf-c-content"
            >
              <dl
                class=""
                data-pf-content="true"
              >
                <h2
                  class=""
                  data-pf-content="true"
                >
                  Copy
                </h2>
                <dt
                  class=""
                  data-pf-content="true"
                >
                  Ctrl + C
                </dt>
                <dd
                  class=""
                  data-pf-content="true"
                />
                <h2
                  class=""
                  data-pf-content="true"
                >
                  Help
                </h2>
                <dt
                  class=""
                  data-pf-content="true"
                >
                  Shift + /
                </dt>
                <dd
                  class=""
                  data-pf-content="true"
                >
                  Show keyboard shortcuts
                </dd>
              </dl>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
